Add Cloud Events support to Spring Integration

cppwfs opened the pull request:
Introduces Cloud Events v1.0 specification support including message converters, transformers, and utilities.

Key components added:
- CloudEventMessageConverter for message format conversion
- ToCloudEventTransformer for transforming messages to Cloud Events
- MessageBinaryMessageReader/Writer for binary format handling
- CloudEventProperties for configuration management
- Header pattern matching utilities for flexible event mapping
- Add reference docs and what's-new paragraph

<!--
Thanks for contributing to Spring Integration. 
Please provide a brief description of your pull-request and reference any related issue numbers (prefix references with #) or StackOverflow questions.

See the [Contributor Guidelines for more information](https://github.com/spring-projects/spring-integration/blob/main/CONTRIBUTING.adoc).
-->

cppwfs committed:
Add Cloud Events support to Spring Integration

Introduces Cloud Events v1.0 specification support
including message converters, transformers, and utilities.

Key components added:
- CloudEventMessageConverter for message format conversion
- ToCloudEventTransformer for transforming messages to Cloud Events
- MessageBinaryMessageReader/Writer for binary format handling
- CloudEventProperties for configuration management
- Header pattern matching utilities for flexible event mapping
- Add reference docs and what's-new paragraph

artembilan commented on build.gradle:
```diff
@@ -472,6 +473,24 @@ project('spring-integration-cassandra') {
 }
 
+project('spring-integration-cloudevents') {
+    description = 'Spring Integration Cloud Events Support'
```
`CloudEvents` ?

artembilan commented on build.gradle:
```diff
@@ -472,6 +473,24 @@ project('spring-integration-cassandra') {
+        }
+        optionalApi "org.apache.avro:avro:$avroVersion"
+        optionalApi "io.cloudevents:cloudevents-xml:$cloudEventsVersion"
+        optionalApi 'org.jspecify:jspecify'
```
I don't think this is something we need to specify manually.
It has to come from Spring Core.

artembilan commented on build.gradle:
```diff
@@ -472,6 +473,24 @@ project('spring-integration-cassandra') {
+            exclude group: 'org.apache.avro', module: 'avro'
+        }
+        optionalApi "org.apache.avro:avro:$avroVersion"
+        optionalApi "io.cloudevents:cloudevents-xml:$cloudEventsVersion"
```
Are these all `optional` deps really necessary for our production code?
Maybe `test` scope instead?

artembilan commented on build.gradle:
```diff
@@ -472,6 +473,24 @@ project('spring-integration-cassandra') {
+        optionalApi "io.cloudevents:cloudevents-xml:$cloudEventsVersion"
+        optionalApi 'org.jspecify:jspecify'
+
+        testImplementation 'org.springframework.amqp:spring-rabbit-test'
```
Do we really need to test this stuff against RabbitMQ ?

artembilan commented on spring-integration-cloudevents/src/main/java/org/springframework/integration/cloudevents/v1/CloudEventMessageConverter.java:
```diff
@@ -0,0 +1,112 @@
+ * limitations under the License.
+ */
+
+package org.springframework.integration.cloudevents.v1;
```
I don't think we need a specific `v1` package.
How about just aim for v1 as is in the root package?

artembilan commented on spring-integration-cloudevents/src/main/java/org/springframework/integration/cloudevents/v1/CloudEventMessageConverter.java:
```diff
@@ -0,0 +1,112 @@
+ */
+public class CloudEventMessageConverter implements MessageConverter {
+
+	private String cePrefix;
```
`final`

artembilan commented on spring-integration-cloudevents/src/main/java/org/springframework/integration/cloudevents/v1/CloudEventMessageConverter.java:
```diff
@@ -0,0 +1,112 @@
+
+	@Override
+	public Object fromMessage(Message<?> message, Class<?> targetClass) {
+		Assert.state(CloudEvent.class.isAssignableFrom(targetClass), "Target class must be a CloudEvent");
```
I think we can just ignore the class and state that in the Javadocs.

artembilan commented on spring-integration-cloudevents/src/test/java/org/springframework/integration/cloudevents/v1/transformer/CloudEventMessageConverterTest.java:
```diff
@@ -0,0 +1,242 @@
+import static org.assertj.core.api.Assertions.assertThatIllegalStateException;
+import static org.assertj.core.api.Assertions.catchIllegalStateException;
+
+public class CloudEventMessageConverterTest {
```
`..Tests`, please.

artembilan commented on src/reference/antora/modules/ROOT/pages/whats-new.adoc:
```diff
@@ -112,3 +112,9 @@ See xref:file/remote-persistent-flf.adoc[Remote Persistent File List Filters] fo
+
+[[x7.0-cloudevents]]
+=== CloudEvents
+The CloudEvent transformer converts Spring Integration messages into CloudEvent compliant messages.
```
Blank line after title.

artembilan commented on src/reference/antora/modules/ROOT/pages/whats-new.adoc:
```diff
@@ -112,3 +112,9 @@ See xref:file/remote-persistent-flf.adoc[Remote Persistent File List Filters] fo
+=== CloudEvents
+The CloudEvent transformer converts Spring Integration messages into CloudEvent compliant messages.
+This transformer provides  support for the CloudEvents specification v1.0 with configurable output formats, header pattern matching, and extension management.
+See xref:cloudevents/cloudevents-transform.adoc[] for more information.
```
Why it cannot be just top-level `cloudevents.adoc` ?

artembilan reviewed (changes requested):
Left some at a glance review.

Thank you!

artembilan reviewed (changes requested):
Sorry for a lengthy review and some doubts I've expressed.

I addition, what are your thoughts about content of this package in Spring Cloud Function: https://github.com/spring-cloud/spring-cloud-function/tree/main/spring-cloud-function-context/src/main/java/org/springframework/cloud/function/cloudevent ?

I mean transformer does the trick indeed, but only from an integration flow context.
How about the way to be able to construct `CloudEvent` programmatically?
Or just existing SDK API is enough to deal with?

cppwfs committed:
Refactor CloudEvents package structure

Remove v1 subpackage and flatten the CloudEvents package hierarchy.
Introduce strategy pattern for format conversion to replace enum-based approach,
improving extensibility and reduce dependencies.

Key changes:
- Move all classes from cloudevents.v1 to cloudevents base package
- Remove optional format dependencies (JSON, XML, Avro) from build
- Replace `ConversionType` enum with `FormatStrategy` interface
- Add `CloudEventMessageFormatStrategy` as default implementation
- Inline `HeaderPatternMatcher` logic into `ToCloudEventTransformerExtensions`
- Add `@NullMarked` package annotations and `@Nullable` throughout
- Document `targetClass` parameter behavior in `CloudEventMessageConverter`
- Split transformer tests for better organization and coverage
- Update component type identifier to "ce:to-cloudevents-transformer"
- Remove unnecessary docs from package-info

cppwfs commented on build.gradle:
```diff
@@ -472,6 +473,14 @@ project('spring-integration-cassandra') {
+    description = 'Spring Integration CloudEvents Support'
+
+    dependencies {
+        api "io.cloudevents:cloudevents-core:$cloudEventsVersion"
```
I'm not sure I understand the squash.  This was the first commit following the first review, IIRC, but I could be wrong.

cppwfs committed:
Consolidate CloudEvent transformer configuration

- Simplify the CloudEvent transformer by consolidating configuration
  directly into ToCloudEventTransformer class rather than using
  separate configuration objects
- Remove CloudEventProperties and ToCloudEventTransformerExtensions
  classes to reduce abstraction layers and improve maintainability
- Make MessageBinaryMessageReader package-private and convert
  CloudEventMessageConverter methods to static where possible
- Move extension filtering logic into a private inner class within
  the transformer
- Remove CloudEventsHeaders class and CE_PREFIX constant as the
  prefix is no longer used as a configurable value

cppwfs commented on spring-integration-cloudevents/src/main/java/org/springframework/integration/cloudevents/MessageBuilderMessageWriter.java:
```diff
@@ -0,0 +1,87 @@
+
+	@Override
+	public Message<byte[]> end() {
+		return MessageBuilder.withPayload(new byte[0]).copyHeaders(this.headers).build();
```
Not exactly.   In this implementation the `copyHeaders` is called while in the `end(null)` does not.

cppwfs commented on spring-integration-cloudevents/src/main/java/org/springframework/integration/cloudevents/transformer/strategies/FormatStrategy.java:
```diff
@@ -0,0 +1,44 @@
+	 * @param messageHeaders the headers associated with the {@link Message}
+	 * @return the converted {@link Message}
+	 */
+	Message<?> convert(CloudEvent cloudEvent, MessageHeaders messageHeaders);
```
Let's discuss.    I'm not sure I understand.

cppwfs committed:
Simplify CloudEvent formatting with SDK strategies

Replace custom CloudEvent converter infrastructure with direct
CloudEvents SDK format implementations.

Key changes:
- Replace `FormatStrategy` pattern-based approach with direct
  `EventFormatProvider` integration from CloudEvents SDK
- Remove custom converter classes (`CloudEventMessageConverter`,
  `MessageBinaryMessageReader`, `MessageBuilderMessageWriter`)
- Simplify transformer to use Expression-based configuration for
  all CloudEvent attributes (id, source, type, dataSchema, subject)
- Add validation for required CloudEvent attributes with clear
  error messages when expressions evaluate to null or empty values
- Update documentation to reflect Expression-based API and
  byte[] payload requirement
- Consolidate tests by removing coverage for deleted converter
  infrastructure

artembilan commented on build.gradle:
```diff
@@ -474,6 +475,23 @@ project('spring-integration-cassandra') {
+        optionalApi("io.cloudevents:cloudevents-avro-compact:$cloudEventsVersion") {
+            exclude group: 'org.apache.avro', module: 'avro'
+        }
+        optionalApi "org.apache.avro:avro:$avroVersion"
```
I think we ca rely on whatever their module give us.
More over, I don't see a reason to have formats as compile dependencies at all.
Just `testImplementation`  should be enough for our use-case.
We really must not have any usage of their format implementations in our prod code.

artembilan commented on build.gradle:
```diff
@@ -474,6 +475,23 @@ project('spring-integration-cassandra') {
+        }
+        optionalApi "org.apache.avro:avro:$avroVersion"
+        optionalApi "io.cloudevents:cloudevents-xml:$cloudEventsVersion"
+        testImplementation 'tools.jackson.core:jackson-databind'
```
Why do we need this?
I don't believe that `cloudevents-json-jackson` can work with Jackson 3.
So, just rely on whatever is their transitive dependency.

cppwfs replied:
Removed.   Artifact from a previous test.

artembilan commented on spring-integration-cloudevents/src/main/java/org/springframework/integration/cloudevents/transformer/ToCloudEventTransformer.java:
```diff
@@ -0,0 +1,288 @@
+		}
+		if (this.dataSchemaExpression == null) { // in the case the user sets the value prior to onInit.
+			this.dataSchemaExpression = new FunctionExpression<>((Function<Message<?>, @Nullable URI>)
+					message -> null);
```
Why cannot we deal with just null for this `dataSchemaExpression`?
No need in a `Function` overhead just to return `null`

artembilan commented on spring-integration-cloudevents/src/main/java/org/springframework/integration/cloudevents/transformer/ToCloudEventTransformer.java:
```diff
@@ -0,0 +1,288 @@
+
+	/**
+	 * Transform the input message into a CloudEvent message.
+	 *
```
Something is off with your IDE formatting: no blank lines in method Javadocs.

artembilan commented on spring-integration-cloudevents/src/main/java/org/springframework/integration/cloudevents/transformer/ToCloudEventTransformer.java:
```diff
@@ -0,0 +1,288 @@
+
+		return MessageBuilder.withPayload(eventFormat.serialize(cloudEvent))
+				.copyHeaders(message.getHeaders())
+				.build();
```
We don't need to build a `Message` here. 
See `AbstractTransformer.transform()` code.

artembilan commented on spring-integration-cloudevents/src/main/java/org/springframework/integration/cloudevents/transformer/ToCloudEventTransformer.java:
```diff
@@ -0,0 +1,288 @@
+		@SuppressWarnings("unchecked")
+		ToCloudEventTransformerExtensions(EvaluationContext evaluationContext, Message<byte[]> message,
+				Expression @Nullable ... expressions) {
+			this.cloudEventExtensions = new HashMap<>();
```
I prefer to have a blank line in the method body if header is multi-line.
Much easier to read the code.

artembilan commented on src/reference/antora/modules/ROOT/nav.adoc:
```diff
@@ -124,6 +124,7 @@
 ** xref:amqp/amqp-1.0.adoc[]
 * xref:camel.adoc[]
 * xref:cassandra.adoc[]
+* xref:cloudevents-transform.adoc[]
```
Not a fact that it always be only for transformers.
Please, rename the file to `cloudevents`.

artembilan commented on src/reference/antora/modules/ROOT/pages/cloudevents-transform.adoc:
```diff
@@ -0,0 +1,133 @@
+[[cloudevent-transformer-overview]]
+=== Overview
+
+The CloudEvent transformer (`ToCloudEventTransformer`) extends Spring Integration's `AbstractTransformer` to convert messages to CloudEvent format.
```
No need to say that `AbstractTransformer` is a Spring Integration class.
We are already here.

artembilan commented on src/reference/antora/modules/ROOT/pages/cloudevents-transform.adoc:
```diff
@@ -0,0 +1,133 @@
+----
+
+==== Extension Expressions
+The expressions constructor parameter is an array of `Expression`s.
```
Again: blank line after the title.
And probably this section would be changed if we come back to patterns for headers to map into extensions.
Unless you can convince me why we have to go with expressions for them.

artembilan commented on src/reference/antora/modules/ROOT/pages/cloudevents-transform.adoc:
```diff
@@ -0,0 +1,133 @@
+
+==== Extension Expressions
+The expressions constructor parameter is an array of `Expression`s.
+If the array is null, then no extensions will be added to the CloudEvent.
```
The code has to be back-ticked.

artembilan commented on src/reference/antora/modules/ROOT/pages/cloudevents-transform.adoc:
```diff
@@ -0,0 +1,133 @@
+==== Extension Expressions
+The expressions constructor parameter is an array of `Expression`s.
+If the array is null, then no extensions will be added to the CloudEvent.
+Each `Expression` in the array must return the type Map<String, Object>.
```
DITTO

artembilan commented on src/reference/antora/modules/ROOT/pages/cloudevents-transform.adoc:
```diff
@@ -0,0 +1,133 @@
+Each `Expression` in the array must return the type Map<String, Object>.
+Where the key is a string and the value is of type object.
+In the example below the extensions are hard coded to return 3 `Map<String, Object>` objects each containing one extension.
+[source,java]
```
Blank line before code block: much easier to read the doc.

artembilan commented on src/reference/antora/modules/ROOT/pages/cloudevents-transform.adoc:
```diff
@@ -0,0 +1,133 @@
+| the id of the message.
+
+| `source`
+| Prefix of "/spring/" followed by the appName a `.` then the name of the transformer's bean.
```
Something is off with the part about `a .`

cppwfs commented on spring-integration-cloudevents/src/main/java/org/springframework/integration/cloudevents/transformer/ToCloudEventTransformer.java:
```diff
@@ -0,0 +1,288 @@
+		this.evaluationContext = ExpressionUtils.createStandardEvaluationContext(getBeanFactory());
+		ApplicationContext applicationContext = getApplicationContext();
+		if (this.sourceExpression == null) {  // in the case the user sets the value prior to onInit.
+			this.sourceExpression = new FunctionExpression<>((Function<Message<?>, URI>) message -> {
```
It needs to be a URI.   I could use SpEL

cppwfs commented on spring-integration-cloudevents/src/main/java/org/springframework/integration/cloudevents/transformer/ToCloudEventTransformer.java:
```diff
@@ -0,0 +1,288 @@
+
+		String id = this.idExpression.getValue(this.evaluationContext, message, String.class);
+		if (!StringUtils.hasText(id)) {
+			throw new MessageTransformationException(message, "No id was found with the specified expression");
```
I will let the builder do the work for testing for null and throwing an exception.   No need to do it twice, since performance is critical

cppwfs commented on spring-integration-cloudevents/src/main/java/org/springframework/integration/cloudevents/transformer/ToCloudEventTransformer.java:
```diff
@@ -0,0 +1,288 @@
+			throw new MessageTransformationException(message, "Missing 'Content-Type' header");
+		}
+
+		EventFormat eventFormat = this.eventFormatProvider.resolveFormat(contentType);
```
I'm not quite clear on this request.  Let's discuss.

cppwfs committed:
Refactor CloudEvent extensions and update javadocs

The previous extension extraction mechanism using `Expression` arrays
and a separate `ExtensionsExtractor` interface was overly complex for
the simple use case of pattern matching header names.

This change simplifies the API by:
- Removing the `ExtensionsExtractor` interface and implementations
- Replacing `Expression`-based extension configuration with simple
  String pattern matching in the transformer constructor
- Updating all javadocs to use imperative voice per Spring
  conventions (e.g., "Converts messages" instead of "A transformer
  that converts messages")
- Making default value descriptions more concise (e.g., "Defaults
  to null" instead of "Default Expression evaluates to a null")
- Add extensionPattern match logic

cppwfs committed:
Enable CloudEvents to be sent with headers

Enable CloudEvents to be sent with headers
instead of requiring structured format serialization. This provides
flexibility when integrating with systems that don't support
CloudEvents structured formats.

Introduce `CloudEventMessageConverter` to handle CloudEvent to
Message conversion, utilizing the CloudEvents SDK's
`MessageWriter` abstraction.

Add `noFormat` configuration option to `ToCloudEventsTransformer`.
When enabled and no `EventFormat` is available for the content type,
CloudEvent attributes are written to message headers with
configurable prefix (defaults to "ce-").

Add `cloudEventPrefix` property to customize the header prefix
when `noFormat` is set to true, supporting different integration scenarios.

Add test coverage for binary content mode including
extension handling, custom prefixes, and validation that original
headers are preserved alongside CloudEvent headers.

cppwfs committed:
Improve CloudEvents documentation and javadocs

Update javadoc comments in `CloudEventMessageConverter` and
`ToCloudEventsTransformer` to improve clarity and readability.
Enhance the CloudEvents reference documentation to better explain
how extensions are populated using pattern matching instead of
SpEL expressions.

Changes include:
- Clarify `CloudEventMessageConverter` class-level javadoc
- Improve `isNoFormat()` and `setNoFormat()` method documentation
- Update reference docs to reflect extension patterns approach
- Fix minor formatting issues (double spaces)

artembilan commented on spring-integration-cloudevents/src/main/java/org/springframework/integration/cloudevents/transformer/ToCloudEventsTransformer.java:
```diff
@@ -0,0 +1,323 @@
+	 * Defaults to null.
+	 * @param subjectExpression the expression to create the subject for each CloudEvent
+	 */
+	public void setSubjectExpression(@Nullable Expression subjectExpression) {
```
Why do we allow this and the one before expression to be set as `null`?

artembilan commented on spring-integration-cloudevents/src/main/java/org/springframework/integration/cloudevents/transformer/ToCloudEventsTransformer.java:
```diff
@@ -0,0 +1,323 @@
+ * limitations under the License.
+ */
+
+
```
That's too many blank lines.
Not sure how Checkstyle let fall though, but only one bank line after copyright.

artembilan commented on spring-integration-cloudevents/src/main/java/org/springframework/integration/cloudevents/transformer/ToCloudEventsTransformer.java:
```diff
@@ -0,0 +1,323 @@
+
+		String contentType = message.getHeaders().get(MessageHeaders.CONTENT_TYPE, String.class);
+		if (contentType == null) {
+			throw new MessageTransformationException(message, "Missing 'Content-Type' header");
```
I think we should fallback to the `application/octet-stream` since we have already checked that payload is a `byte[]`.

artembilan commented on spring-integration-cloudevents/src/main/java/org/springframework/integration/cloudevents/transformer/ToCloudEventsTransformer.java:
```diff
@@ -0,0 +1,323 @@
+
+	@Override
+	public String getComponentType() {
+		return "ce:to-cloudevents-transformer";
```
If we rename the class, this value should be revised respectively.

artembilan commented on spring-integration-cloudevents/src/main/java/org/springframework/integration/cloudevents/transformer/ToCloudEventsTransformer.java:
```diff
@@ -0,0 +1,323 @@
+
+		/**
+		 * Construct CloudEvent extensions by processing a message using expressions.
+		 *
```
No blank lines in method Javadoc.

artembilan commented on src/reference/antora/modules/ROOT/pages/cloudevents.adoc:
```diff
@@ -0,0 +1,268 @@
+[[cloudevents]]
+= CloudEvents Support
+
+Spring Integration provides transformers (starting with version 7.0) for transforming messages into CloudEvent messages.
```
The version is out-dated already.

artembilan commented on src/reference/antora/modules/ROOT/pages/cloudevents.adoc:
```diff
@@ -0,0 +1,268 @@
+
+[[cloudevent-transformers]]
+
+== CloudEvent Transformers
```
I don't see a reason in this top-level title.
Why cannot we go immediately to ` ToCloudEventsTransformer`?

artembilan commented on src/reference/antora/modules/ROOT/pages/cloudevents.adoc:
```diff
@@ -0,0 +1,268 @@
+[[cloudevent-transformer-overview]]
+==== Overview
+
+The CloudEvents transformer (`ToCloudEventsTransformer`) converts messages to CloudEvents format.
```
This is a bit misleading when we introduce a `FromClouldEventTransformer`.
Why do we need a separate `Overview` title?

artembilan commented on src/reference/antora/modules/ROOT/pages/cloudevents.adoc:
```diff
@@ -0,0 +1,268 @@
+The `ToCloudEventsTransformer` allows the user to use SpEL `Expression`s to populate the attributes.
+Extensions are populated from the headers using pattern matching.
+
+====== Attribute Expressions
```
Why do we need so many sub-titles?
Why cannot we express everything in the same section, but different paragraphs?

cppwfs replied:
No.  I'll be removing many of these.

artembilan commented on src/reference/antora/modules/ROOT/pages/cloudevents.adoc:
```diff
@@ -0,0 +1,268 @@
+
+[source,java]
+----
+String[] extensionPatterns = {"trace*", "span-id", "user-id"};
```
Why do we need a separate array variable if we can just use varargs?

artembilan commented on src/reference/antora/modules/ROOT/pages/cloudevents.adoc:
```diff
@@ -0,0 +1,268 @@
+
+[[jsonformat]]
+====== JsonFormat
+The following dependency can be used to include this `JsonFormat` in your project.
```
No sticking to the title, please.
And that is, of course, if we move on with "may sub-titles" approach 😄

artembilan reviewed (changes requested):
Sorry for lengthy review: too many concerns.

Please, consider to rebase your branch to the latest `main`.
And take into account that it is already `7.1`.

Thanks

cppwfs committed:
Refactor CloudEvents transformer for version 7.1

This commit updates the CloudEvents module from version 7.0 to 7.1 by
consolidating classes and improving the architecture. The previous
design separated the CloudEvent conversion logic across multiple
classes (`CloudEventMessageConverter` and `MessageBuilderMessageWriter`),
which added unnecessary complexity for what is fundamentally a single
transformation operation.

Key changes:
- Removed `CloudEventMessageConverter` and `MessageBuilderMessageWriter`
  as standalone classes
- Consolidated conversion logic into `ToCloudEventTransformer` by
  making `MessageBuilderMessageWriter` a private static inner class
- Renamed `ToCloudEventsTransformer` to `ToCloudEventTransformer` to
  match naming conventions (singular form)
- Removed Avro dependencies and associated tests that were no longer
  needed
- Changed `setNoFormat()` to `setFailOnNoFormat()` for clearer semantics
- Enhanced null safety by removing `@Nullable` from setters that should
  always have values
- Re-add isFailOnNoFormat logic and tests that was removed
- Improved documentation and added defensive copying for the
  `extensionPatterns` array parameter
- Updated tests to reflect the new class names and structure
- Rebased

The refactoring reduces the public API surface while maintaining all
functionality, making the code easier to understand and maintain.

cppwfs commented on spring-integration-cloudevents/src/main/java/org/springframework/integration/cloudevents/transformer/MessageBuilderMessageWriter.java:
```diff
@@ -0,0 +1,137 @@
+	 * @param dataContentTypeKey the header name for the data content type
+	 * @param headers the base message headers to include in the output message
+	 */
+	MessageBuilderMessageWriter(String cloudEventPrefix, String specVersionKey, String dataContentTypeKey, Map<String, Object> headers) {
```
I understood that as a requirement for the transformer.  I am willing to drop it.

cppwfs commented on spring-integration-cloudevents/src/main/java/org/springframework/integration/cloudevents/transformer/ToCloudEventsTransformer.java:
```diff
@@ -0,0 +1,323 @@
+ *
+ * @since 7.0
+ */
+public class ToCloudEventsTransformer extends AbstractTransformer {
```
True.   It was named after the `cloudevents` project.

artembilan commented on spring-integration-cloudevents/src/main/java/org/springframework/integration/cloudevents/transformer/ToCloudEventsTransformer.java:
```diff
@@ -0,0 +1,323 @@
+	}
+
+	private String getDataContentTypeKey() {
+		return this.cloudEventPrefix + DEFAULT_DATACONTENTTYPE_KEY;
```
That's the point.
We just propagate a `prefix` down to the writer.
The rest is internal logic for prefixing those standard keys.

cppwfs replied:
Writer now only accepts prefix, and it does the construction.   All other mentions of these variables have been removed from the transformer.

cppwfs commented on src/reference/antora/modules/ROOT/pages/cloudevents.adoc:
```diff
@@ -0,0 +1,268 @@
+----
+compile "io.cloudevents:cloudevents-json-jackson:$cloudEventsVersion"
+----
+======
```
In this case I removed them completely.    Simplifying our reference docs, and sourcing the sdk's reference docs.

artembilan reviewed (changes requested):
I see you have addressed everything from the previous review, but probably just locally.
Please, push it into PR with rebase to the latest `main`.
This way I might review it again before Monday.

Thanks

cppwfs committed:
Refine CloudEvents transformer implementation

Address multiple improvements to the CloudEvents transformer module:

- Reorganize constructor ordering in `ToCloudEventTransformer` to put
  the no-arg constructor first, following Java conventions
- Remove unused Avro test dependency from build.gradle
- Improve content type handling using `MessageHeaderAccessor`
  and `MimeType` instead of direct String extraction
- Refactor extension processing to use `Map.copyOf()` for immutability
- Rename `ToCloudEventTransformerExtensions` to singular
  `ToCloudEventTransformerExtension` for consistency
- Inline constant strings for specversion and datacontenttype keys
- Fix content type header to use `eventFormat.serializedContentType()`
  instead of the input message's content type
- Simplify binary mode message creation by constructing
  `MessageHeaders` directly
- Update test assertions to match null appName behavior
- Enhance documentation clarity: improve wording, fix grammar,
  capitalize table entries, and add examples for content types
- Restructure documentation sections for better readability
- Update `MessageBuilderMessageWriter` to implement
  `CloudEventWriterFactory` instead of `MessageWriter`
  The message writer had default functions unused by the
  read method.
- Remove unused `setEvent()` method and `dataContentTypeKey` field
  from `MessageBuilderMessageWriter`
- Apply Spring code convention

artembilan commented on spring-integration-cloudevents/src/main/java/org/springframework/integration/cloudevents/transformer/ToCloudEventTransformer.java:
```diff
@@ -0,0 +1,415 @@
+	protected void onInit() {
+		super.onInit();
+		this.evaluationContext = ExpressionUtils.createStandardEvaluationContext(getBeanFactory());
+		if (this.sourceExpression == null) {  // in the case the user sets the value prior to onInit.
```
Please, remove the comment.
It does not give any value. 
Plus, it sounds like does not reflect reality: the condition is about `null`, but sentence in comment says something opposite.

artembilan commented on spring-integration-cloudevents/src/main/java/org/springframework/integration/cloudevents/transformer/ToCloudEventTransformer.java:
```diff
@@ -0,0 +1,415 @@
+		String type = this.typeExpression.getValue(this.evaluationContext, message, String.class);
+		MessageHeaders headers = message.getHeaders();
+		MessageHeaderAccessor accessor = new MessageHeaderAccessor(message);
+		MimeType mimeType = accessor.getContentType();
```
Please, consider to use `StaticMessageHeaderAccessor.getContentType()` instead.
That way we won't have an extra `MessageHeaderAccessor` volatile object.

artembilan commented on spring-integration-cloudevents/src/main/java/org/springframework/integration/cloudevents/transformer/ToCloudEventTransformer.java:
```diff
@@ -0,0 +1,415 @@
+		 */
+		@Override
+		public Message<byte[]> end() {
+			return org.springframework.integration.support.MessageBuilder
```
Why do we need to use fully qualified class name?

artembilan commented on spring-integration-cloudevents/src/main/java/org/springframework/integration/cloudevents/transformer/ToCloudEventTransformer.java:
```diff
@@ -0,0 +1,415 @@
+		 * Create a MessageBuilderMessageWriter with the specified configuration.
+		 * @param cloudEventPrefix the prefix to prepend to CloudEvent attribute names in message headers
+		 * @param headers the base message headers to include in the output message
+		 */
```
Same here: Javadoc on the `MessageBuilderMessageWriter` itself, not on the private ctor.

artembilan commented on spring-integration-cloudevents/src/test/java/org/springframework/integration/cloudevents/transformer/ToCloudEventTransformerTests.java:
```diff
@@ -0,0 +1,330 @@
+	@Test
+	void failWhenNoIdHeaderAndNoDefault() {
+		Message<byte[]> message = MessageBuilder.withPayload(PAYLOAD)
+				.setHeader("contentType", JsonFormat.CONTENT_TYPE)
```
Use a convenient `MessageHeaders.CONTENT_TYPE` instead.
And why then do you have `createBaseMessage()` if you don't use it?

artembilan commented on spring-integration-cloudevents/src/test/java/org/springframework/integration/cloudevents/transformer/ToCloudEventTransformerTests.java:
```diff
@@ -0,0 +1,330 @@
+		Object result = this.transformerWithNoExtensions.doTransform(message);
+
+		assertThat(result).isNotNull();
+		assertThat(result).isInstanceOf(Message.class);
```
The `isNotNull()` is redundant.
The AsssertJ is smart enough to cover us with.
Plus, it is a fluent API, so we don't need to call `assertThat()` on the same object again: just chain assertions.

artembilan commented on spring-integration-cloudevents/src/test/java/org/springframework/integration/cloudevents/transformer/ToCloudEventTransformerTests.java:
```diff
@@ -0,0 +1,330 @@
+		String payload = "test message";
+		Message<byte[]> message = createBaseMessage(payload.getBytes(), "application/cloudevents+json")
+			.setHeader("correlation-id", "corr-999")
+				.setHeader(TRACE_HEADER, "trace-123")
```
Something is off with indents on these chained methods.

artembilan commented on spring-integration-cloudevents/src/test/java/org/springframework/integration/cloudevents/transformer/ToCloudEventTransformerTests.java:
```diff
@@ -0,0 +1,330 @@
+
+	private static final String USER_HEADER = "userId";
+
+	private static final String[] TEST_PATTERNS = {"trace*", SPAN_HEADER, USER_HEADER};
```
This property is not used in the tests, so consider to move it into a `@Configuration` class for a legit place of the responsibility around it.

artembilan commented on spring-integration-cloudevents/src/main/java/org/springframework/integration/cloudevents/transformer/ToCloudEventTransformer.java:
```diff
@@ -0,0 +1,415 @@
+ * Perform attribute and extension mapping based on {@link Expression}s.
+ *
+ * @author Glenn Renfro
+ * @since 7.1
```
I prefer to put a blank line after the `@author`, so it becomes more readable who got a credit.

artembilan commented on spring-integration-cloudevents/src/main/java/org/springframework/integration/cloudevents/transformer/package-info.java:
```diff
@@ -0,0 +1,3 @@
+
+@org.jspecify.annotations.NullMarked
```
Javadoc is missed in this file.

artembilan commented on spring-integration-cloudevents/src/main/java/org/springframework/integration/cloudevents/package-info.java:
```diff
@@ -0,0 +1,3 @@
+
+@org.jspecify.annotations.NullMarked
```
Javadoc is missed in this file.
That gives a nice summary in the list of packages: https://docs.spring.io/spring-integration/docs/7.0.1/api/

artembilan commented on spring-integration-cloudevents/src/test/java/org/springframework/integration/cloudevents/transformer/ToCloudEventTransformerTests.java:
```diff
@@ -0,0 +1,330 @@
+				.setHeader(MessageHeaders.CONTENT_TYPE, contentType);
+	}
+
+	@Configuration
```
Consider to use `proxyBeanMethods = false` for slight performance improvement.

artembilan commented on spring-integration-cloudevents/src/test/java/org/springframework/integration/cloudevents/transformer/ToCloudEventTransformerTests.java:
```diff
@@ -0,0 +1,330 @@
+
+		@Bean
+		public ToCloudEventTransformer transformerWithNoExtensionsNoFormatEnabled() {
+			ToCloudEventTransformer toCloudEventsTransformer =  new ToCloudEventTransformer();
```
Please, do some reformatting with your IDE: all those extra spaces should go away.

artembilan commented on src/reference/antora/modules/ROOT/pages/cloudevents.adoc:
```diff
@@ -0,0 +1,135 @@
+    .withHeader(MessageHeaders.CONTENT_TYPE, "application/octet-stream")
+    .build();
+// Transformer with extension patterns
+ToCloudEventTransformer transformer = new ToCloudEventTransformer();
```
The comment about does not reflect this constructor behavior.

artembilan reviewed (changes requested):
Sorry for such a lengthy review before New Year 😄 .

cppwfs committed:
Refine CloudEvents transformer code quality and tests

Code quality improvements:
- Remove unnecessary `@since 7.1` tags from internal javadocs
- Fix NullAway suppressions by adding proper null checks
- Replace `MessageHeaderAccessor` with `StaticMessageHeaderAccessor`
- Eliminate redundant field initialization (failOnNoFormat defaults to false)
- Simplify extension comparison using `Boolean.TRUE.equals()`
- Remove unnecessary Map.copyOf() in extension implementation
- Fix app name null check to only warn when actually null

Test refactoring:
- Make test helper methods static for better encapsulation
- Rename test methods to clearly indicate content type being tested
- Separate JSON format vs application/json content type test cases
- Add explicit verification for application/* content types
- Improve test assertions using containsEntry() instead of get()
- Remove unnecessary null checks in tests
- Consolidate duplicate test setup code

Documentation:
- Add package-info.java descriptions for cloudevents packages
- Update reference documentation for extension patterns clarity

artembilan commented on spring-integration-cloudevents/src/test/java/org/springframework/integration/cloudevents/transformer/ToCloudEventTransformerTests.java:
```diff
@@ -0,0 +1,330 @@
+
+	@Test
+	void emptyStringPayloadHandling() {
+		Message<byte[]> message = createBaseMessage("".getBytes(), "application/cloudevents+json").build();
```
Just to be sure I understood that right.
Do you mean that this test is about receiving a cloud event message and produce another cloud event message?
Then that is not clear from the test name and its body.
Plus the logic of our transformer is not aware of the cloud event as an input and we just create an new one with the body of the previous one.
That is not just adding extensions to existing one as you claim.
At least that is what I see from our logic so far.

At the same time another question.
If you use `application/cloudevents+json` explicitly then you mean that `JsonFormat.CONTENT_TYPE` is just `application/json`  ?
And to be more clear: do we have test coverage where an input message has an `application/json` as its `contentType` ?

artembilan replied:
Alright.
Taking that back.
I see what is going on.
Their `EventFormatProvider` indeed expects from us something like `application/cloudevents+json`.
https://github.com/cloudevents/sdk-java/blob/main/core/src/main/java/io/cloudevents/core/provider/EventFormatProvider.java#L71
which comes from the: https://github.com/cloudevents/sdk-java/blob/main/formats/json-jackson/src/main/java/io/cloudevents/jackson/JsonFormat.java#L181 or https://github.com/cloudevents/sdk-java/blob/main/formats/avro-compact/src/main/java/io/cloudevents/avro/compact/AvroCompactFormat.java#L132.

So, your current logic is correct. 
According to our discussion and decision around `contentType` header it is really correct.
It is not what would be great to have in the end as convenient solution, but for now let's leave it as is and come back to it after New Year.
Need more thinking how to map normal `text/json` into their `EventFormatProvider` expectations.
If that is necessary at all, though...

cppwfs commented on spring-integration-cloudevents/src/test/java/org/springframework/integration/cloudevents/transformer/ToCloudEventTransformerTests.java:
```diff
@@ -0,0 +1,330 @@
+		assertThat(cloudEvent.getData().toBytes()).isEqualTo(PAYLOAD);
+		assertThat(cloudEvent.getSource().toString()).isEqualTo("/spring/null.transformerWithNoExtensions");
+		assertThat(cloudEvent.getDataSchema()).isNull();
+		assertThat(cloudEvent.getDataContentType()).isEqualTo(JsonFormat.CONTENT_TYPE);
```
Based on our earlier exchange I think this question has been answered.  If not please let me know.

artembilan commented on spring-integration-cloudevents/src/main/java/org/springframework/integration/cloudevents/transformer/ToCloudEventTransformer.java:
```diff
@@ -0,0 +1,404 @@
+
+/**
+ * Convert messages to CloudEvent format.
+ * Perform attribute and extension mapping based on {@link Expression}s.
```
OK with `attribute`.
But `extension` is not does by expression.
Please, consider to revise this Javadoc to reflect reality about the transformer logic.
I guess we can leave `EventFormat` out of this for now, since we don't have a final solution yet.
But `extensionPatterns` could be already explained in this Javadocs properly.

cppwfs committed:
Use injection and expressions to set EventFormat

Key changes:
- Add eventFormatContentTypeExpression as a method for a user to use the EventFormatProvider
- Allow user to inject EventFormat into the transformer
- Resolved contentType resolution allowing the eventFormat or the MessageBuilderMessageWriter to determine it.
- Add constructors accepting `@Nullable` `EventFormat` parameter for
  dependency injection, enabling flexible format configuration
- Refactor default constructor to delegate to parameterized
  constructor, avoiding code duplication
- Add `eventFormatContentTypeExpression` property to support dynamic
  format resolution via `EventFormatProvider`
- Reorder fields per Spring conventions (static fields first)
- Update javadocs to use imperative style and improve clarity
- Simplify `getCloudEventExtensions()` with inline declarations
- Remove redundant Content-Type header assignment in binary mode
- Update test configuration to inject `EventFormat` instances
- Add test coverage for `EventFormatProvider` expression usage
- Refine test assertions to verify CloudEvent serialization
- Update reference documentation with latest changes
- Rebased

artembilan commented on spring-integration-cloudevents/src/main/java/org/springframework/integration/cloudevents/transformer/ToCloudEventTransformer.java:
```diff
@@ -0,0 +1,426 @@
+	}
+
+	/**
+	 * Set the {@link Expression} to create CloudEvent ids.
```
Just `id`.
I think talking here in the context of one `CloudEvent` instance is OK.
Therefore using `ids` word is misleading a bit.

artembilan commented on src/reference/antora/modules/ROOT/pages/cloudevents.adoc:
```diff
@@ -0,0 +1,171 @@
+[[cloudevents]]
+= CloudEvents Support
+
+Use Spring Integration transformers (starting with version 7.1) to transform messages into CloudEvent messages.
```
I think the general sentence in this chapter should talk about a general Cloud Events support in Spring Integration.
The fact that we have just a transformer (for now) must not be exposed so explicitly.
We talk about transformer in its dedicated sub-section.

artembilan commented on src/reference/antora/modules/ROOT/pages/cloudevents.adoc:
```diff
@@ -0,0 +1,171 @@
+
+==== Attribute Expressions
+
+Set the CloudEvents' attributes of `id`, `source`, `type`, `dataSchema`, and `subject` through SpEL ``Expression``s.
```
Just plain word `expressions` should be enough.
I mean no need to wrap it into a code.

artembilan commented on src/reference/antora/modules/ROOT/pages/cloudevents.adoc:
```diff
@@ -0,0 +1,171 @@
+
+NOTE: The transformer sets the `time` attribute to the time when it creates the CloudEvent message.
+
+The following table lists the attribute names and the values the default ``Expression``s return.
```
Just `expressions`

artembilan commented on src/reference/antora/modules/ROOT/pages/cloudevents.adoc:
```diff
@@ -0,0 +1,171 @@
+| `null`
+
+| `time`
+| The time the CloudEvent message is created.
```
This should be explained somehow that there is no a dedicated property on the transformer.
And the value is set to the current time internally.

artembilan commented on src/reference/antora/modules/ROOT/pages/cloudevents.adoc:
```diff
@@ -0,0 +1,171 @@
+[[eventformats]]
+
+==== EventFormats
+Use ``EventFormat``s to serialize the CloudEvent into the message's payload when the EventFormat is available, or use Binary Format Mode when an `EventFormat` is not available.
```
Black line after section title.
The `EventFormat` should be a code snippet.
Probably this sentence has to be reworked from the transformer behavior perspective, not an end-user command which is not true in regards of Binary Format since end-user does not chose it. That is a logic of the transformer.

artembilan commented on src/reference/antora/modules/ROOT/pages/cloudevents.adoc:
```diff
@@ -0,0 +1,171 @@
+
+To utilize a specific `EventFormat`, add the associated dependency.
+For example, to add the XML `EventFormat`, add the following dependency: `io.cloudevents:cloudevents-xml`.
+See the https://github.com/cloudevents/sdk-java[CloudEvents SDK project's] reference docs for information on the ``EventFormat``s that are available.
```
We don't use urban language in the official docs.
Therefore, the `reference documentation`  is what we want here instead.

artembilan reviewed (changes requested):
Thank you for the update!
Please, forgive my non-stopping nit-picks.
We are really very close for our never-ending revisions. 😄

cppwfs commented on spring-integration-cloudevents/src/main/java/org/springframework/integration/cloudevents/transformer/ToCloudEventTransformer.java:
```diff
@@ -0,0 +1,426 @@
+	 * set as an output message's payload, and its attributes are set into headers.
+	 * @param failOnNoFormat true to disable format serialization
+	 */
+	public void setFailOnNoFormat(boolean failOnNoFormat) {
```
So are we saying that if they set the `eventFormatContentTypeExpression` and it returns null that we should throw a new `MessageTransformationException`.   
And remove the `setFailOnNoFormat` because we are now throwing an exception in the only path that could lead to an unintentional null for `EventFormat`.  

Just want to clarify.

cppwfs commented on spring-integration-cloudevents/src/main/java/org/springframework/integration/cloudevents/transformer/ToCloudEventTransformer.java:
```diff
@@ -0,0 +1,426 @@
+	 * Return the prefix used for CloudEvent headers in binary content mode.
+	 * @return the CloudEvent header prefix
+	 */
+	public String getCloudEventPrefix() {
```
Wrong selection on the _ctrl-n_ option :-)

artembilan commented on spring-integration-cloudevents/src/main/java/org/springframework/integration/cloudevents/transformer/ToCloudEventTransformer.java:
```diff
@@ -0,0 +1,426 @@
+			contentType = "application/octet-stream";
+		}
+		else {
+			contentType = mimeType.toString();
```
It depends on the length and complexity of the evaluation expression.

cppwfs committed:
Refactor CloudEvents transformer configuration

Remove EventFormat from constructor parameters in favor of setter
injection. This change simplifies the API by eliminating the
`failOnNoFormat` property and consolidating EventFormat configuration
through `setEventFormat()` and `setEventFormatContentTypeExpression()`.

The transformer now throws `MessageTransformationException` only when
`eventFormatContentTypeExpression` is set but cannot resolve to a valid
EventFormat. When neither EventFormat nor expression is specified, the
transformer defaults to Binary Format Mode, writing CloudEvent
attributes to message headers with the configured prefix.

Key changes:
- Remove EventFormat parameter from constructors
- Add `setEventFormat()` method for explicit format configuration
- Simplify `eventFormatContentTypeExpression` error handling
- Remove `failOnNoFormat` property and related getter/setter
- Update test configurations to use setter-based injection
- Reorder MessageBuilderMessageWriter methods for better readability
- Add missing javadocs for CloudEvent attribute mapping methods
- Update documentation to reflect new configuration approach

artembilan commented on spring-integration-cloudevents/src/main/java/org/springframework/integration/cloudevents/transformer/ToCloudEventTransformer.java:
```diff
@@ -0,0 +1,393 @@
+	}
+
+	/**
+	 * Set the {@link Expression} to create the subject for the CloudEvent.
```
`{@code subject}`

artembilan commented on spring-integration-cloudevents/src/main/java/org/springframework/integration/cloudevents/transformer/ToCloudEventTransformer.java:
```diff
@@ -0,0 +1,393 @@
+				.build();
+
+		if (this.eventFormat == null && this.eventFormatContentTypeExpression != null) {
+			this.eventFormat = this.eventFormatProvider.resolveFormat(
```
This is not correct.
You are doing a mutation to the not-thread safe object.
Consider to use a local variable instead of mutating the property of the transformer.

artembilan commented on spring-integration-cloudevents/src/main/java/org/springframework/integration/cloudevents/transformer/ToCloudEventTransformer.java:
```diff
@@ -0,0 +1,393 @@
+					this.eventFormatContentTypeExpression.getValue(this.evaluationContext, message, String.class));
+			if (this.eventFormat == null) {
+				throw new MessageTransformationException("No EventFormat found for the expression '" +
+						this.eventFormatContentTypeExpression.getExpressionString() + "'");
```
Can we use a result of the expression here instead?
That would be much useful to determine what is wrong with the `EventFormatProvider`.
Just talking about an expression does not give us a clue what exactly format content type is missed in the registry.

cppwfs committed:
Update reference to cloudevent jdk

- Replace the word Docs with Documentation

artembilan commented on spring-integration-cloudevents/src/main/java/org/springframework/integration/cloudevents/transformer/ToCloudEventTransformer.java:
```diff
@@ -0,0 +1,393 @@
+import org.springframework.util.StringUtils;
+
+/**
+ * Transform messages to CloudEvent format with attributes and extensions mapping.
```
Would be very useful to have some brief explanation about `EventFormat` and `MessageBuilderMessageWriter`.
And also mention what output message contains in the end, including that specific `MessageHeaders.CONTENT_TYPE`.

artembilan commented on src/reference/antora/modules/ROOT/pages/cloudevents.adoc:
```diff
@@ -0,0 +1,164 @@
+[[cloudevents]]
+= CloudEvents Support
+
+Spring Integration provides support for the CloudEvents specification as discussed https://cloudevents.io/[here].
```
I don't think it is really discussed there.
Why just don't do `https://cloudevents.io[specification]` instead to avoid any possible questions?

artembilan commented on spring-integration-cloudevents/src/test/java/org/springframework/integration/cloudevents/transformer/ToCloudEventTransformerTests.java:
```diff
@@ -0,0 +1,380 @@
+		}
+	}
+
+	private record TestRecord(String sampleValue) implements Serializable { }
```
Looks like we don't do Java serialization any more.

cppwfs committed:
Improve CloudEvents transformer documentation and implementation

Enhance the ToCloudEventTransformer class-level Javadoc to clearly
explain the two output modes: Structured Content Mode and Binary
Content Mode. Document how EventFormat and MessageBuilderMessageWriter
are used, and specify the CONTENT_TYPE header value for each mode.

Refactor the doTransform method to use a local `selectedEventFormat`
variable instead of mutating the instance field, improving thread
safety and clarity. Move the CONTENT_TYPE header setting for binary
mode from the end() methods to the create() method in
MessageBuilderMessageWriter for consistency.

Update Javadoc comments throughout to use `@code` for attribute names
like `dataSchema` and `subject`, and improve clarity in method
descriptions. Refine the precedence explanation between `eventFormat`
and `eventFormatContentTypeExpression` to be more concise. Convert
external URL to proper `@see` link format.

Fix test assertions to expect "application/cloudevents" as the
CONTENT_TYPE in binary mode instead of the original content type.
Update test for serialized objects to use proper Java serialization
via ObjectOutputStream and DefaultSerializer. Improve error message
assertion to include the full expected text with content type details.

Update reference documentation to clarify EventFormat behavior,
improve wording consistency, and fix example code. Add backticks
around module name in whats-new.adoc for proper formatting.

artembilan commented on spring-integration-cloudevents/src/main/java/org/springframework/integration/cloudevents/transformer/ToCloudEventTransformer.java:
```diff
@@ -0,0 +1,406 @@
+	 * content type for the {@link EventFormatProvider}
+	 * More can be read about content types EventFormatProvider supports here:
+	 * @see <a href="
+	 * https://github.com/cloudevents/sdk-java/blob/main/core/src/main/java/io/cloudevents/core/format/ContentType.java"
```
I meant this here:
```
@see io.cloudevents.core.format.ContentType
```
which is natural way to point from one class to another in Javadocs.

artembilan commented on spring-integration-cloudevents/src/test/java/org/springframework/integration/cloudevents/transformer/ToCloudEventTransformerTests.java:
```diff
@@ -0,0 +1,385 @@
+		Message<byte[]> resultMessage = (Message<byte[]>) result;
+		CloudEvent cloudEvent = this.jsonFormat.deserialize(resultMessage.getPayload());
+		assertThat(resultMessage.getHeaders()).containsEntry("correlation-id", "corr-999");
+		assertThat(cloudEvent.getExtensionNames()).containsExactly("traceid").
```
I think that period in the end would look better on the next line next to the method we are calling there.

artembilan commented on src/reference/antora/modules/ROOT/pages/cloudevents.adoc:
```diff
@@ -0,0 +1,164 @@
+======
+
+[[to-cloud-event-transformer]]
+
```
I think there was some misunderstanding.
The section id and its title lines should stick to each other, but the body of the section should be over a blank line after the title.

cppwfs committed:
Update code based on review

Simplify Javadoc `@see` reference to use class name instead of
external URL link for `ContentType` class. This improves readability
and follows Spring Javadoc conventions for referencing types in the
same dependency.

Use the local `selectedEventFormat` variable consistently in the
conditional check instead of checking `this.eventFormat` again. This
improves code clarity and ensures the logic matches the variable
already assigned.

Fix test assertion formatting by moving the chained method call to
the next line, improving readability and following standard fluent
assertion formatting conventions.

artembilan reviewed (changes requested):
Please, mention in the next commit that it is going to fix this issue: https://github.com/spring-projects/spring-integration/issues/3103.
Thanks

artembilan reviewed (approved):
Cool!
Pulling locally for final review and potential merge.
Thank you!

cppwfs commented:
Thank you for all the effort you put into this!   It is sooooo appreciated!
